server/user_activity.go

```diff
@@ -0,0 +1,154 @@
+package main
+
+import (
+	"time"
+
+	"github.com/mattermost/mattermost-server/model"
+)
+
+type userChannelActivity = map[string]map[string]int64 //map[userID]map[channelID]activity
+
+// getActivity returns user activity in channels for every user from the beginning of times.
+// For now user activity is the number of posts posted in a channel.
+func (p *Plugin) getActivity() (userChannelActivity, error) {
```

Comment:
As you note in the comment, this chain of calls amounts to a full query of all posts across all public channels the first time the plugin runs. This will lead to some performance concerns in hitting the database, not to mention memory overhead for the plugin itself -- it may not finish successfully, and be effectively stuck in a crash loop.

Can we find a way to approximate this data instead? A simple idea would be to limit our search to a fixed number of posts. Presumably as the data gets older, the meaning of "activity" in that channel drops.

Reply:
Yeap, this might be a problem for a database. Memory is not an issue here though. I thought of that so I never have all the posts at once in memory. They are aggregated gradually.

Anyway, we surely can approximate data. Our idea is good, we can have a decay parameter which will decrease the post "weight" in time, so older posts will have very little weight. Or we can just get posts from last year(or some other period) it will be easiest. More complicated way will be to get channel posts for some period, get dates when users joined the channel and calculate from there(assuming we have same distribution).
And the right way is to have a DB for data science and machine learning tasks and save aggregated data there :))